common/nexus/payload_serializer.go

```diff
@@ -0,0 +1,221 @@
+	if !ok {
+		return fmt.Errorf("%w: cannot deserialize into %v", errSerializer, v)
+	}
+
+	payload := &commonpb.Payload{}
+	*payloadRef = payload
+	payload.Metadata = make(map[string][]byte)
+	payload.Data = content.Data
+
+	if !isStandardNexusContent(content) {
+		for k, v := range content.Header {
+			payload.Metadata[k] = []byte(v)
+		}
+		payload.Metadata["encoding"] = []byte("unknown/nexus-content")
+		return nil
+	}
+
+	contentType := content.Header.Get("type")
+	if contentType == "" {
+		payload.Metadata["encoding"] = []byte("binary/null")
+		return nil
+	}
+
+	mediaType, params, err := mime.ParseMediaType(contentType)
+	if err != nil {
+		return err
+	}
+
+	switch mediaType {
+	case "application/x-temporal-payload":
+		err := payload.Unmarshal(content.Data)
+		if err != nil {
+			return err
+		}
+		return nil
+	case "application/json":
+		if params["format"] == "protobuf" {
+			payload.Metadata["encoding"] = []byte("json/protobuf")
+			messageType := params["message-type"]
+			if messageType != "" {
+				payload.Metadata["messageType"] = []byte(messageType)
+			}
+		} else {
+			payload.Metadata["encoding"] = []byte("json/plain")
+		}
+	case "application/x-protobuf":
+		payload.Metadata["encoding"] = []byte("binary/protobuf")
+		messageType := params["message-type"]
+		if messageType != "" {
+			payload.Metadata["messageType"] = []byte(messageType)
+		}
+	case "application/octet-stream":
+		payload.Metadata["encoding"] = []byte("binary/plain")
+	default:
+		// Should be unreachable.
+		return fmt.Errorf("%w: standard content detection failed for %q", errSerializer, mediaType)
+	}
+	return nil
+}
+
+func isStandardNexusContent(content *nexus.Content) bool {
+	h := content.Header
+	// We assume that encoding is handled by the transport layer and the content is decoded.
+	delete(h, "encoding")
+	// Length can safely be ignored.
+	delete(h, "length")
+
+	if len(h) > 1 {
+		return false
+	}
+	contentType := h.Get("type")
+	if contentType == "" {
+		return len(h) == 0 && len(content.Data) == 0
+	}
+	mediaType, params, err := mime.ParseMediaType(contentType)
+	if err != nil {
+		return false
+	}
+
+	switch mediaType {
+	case "application/octet-stream",
+		"application/x-temporal-payload":
+		return len(params) == 0
+	case "application/json":
+		if params["format"] == "protobuf" {
+			if params["message-type"] != "" {
+				return len(params) == 2
+			}
+			return len(params) == 1
+		}
+		return len(params) == 0
+	case "application/x-protobuf":
+		if params["message-type"] != "" {
+			return len(params) == 1
+		}
+		return len(params) == 0
+	}
+	return false
+}
+
+// Serialize implements nexus.Serializer.
+func (payloadSerializer) Serialize(v any) (*nexus.Content, error) {
+	payload, ok := v.(*commonpb.Payload)
+	if !ok {
+		return nil, fmt.Errorf("%w: cannot serialize %v", errSerializer, v)
+	}
+
+	if payload == nil {
+		return &nexus.Content{Header: nexus.Header{}}, nil
+	}
+
+	if !isStandardPayload(payload) {
```

Comment:
This is confusing too, can you just make this stuff part of the switch below? You can set data in each arm of the switch, there's not that much code reuse here. Just handle each type specifically. I can suggest exact rewrite if it helps.